jsm.go

```diff
@@ -304,13 +327,21 @@ func (js *js) upsertConsumer(stream string, cfg *ConsumerConfig, opts ...JSOpt)
 	}
 
 	var ccSubj string
-	if cfg != nil && cfg.Durable != _EMPTY_ {
-		if err := checkConsumerName(cfg.Durable); err != nil {
-			return nil, err
+	if consumerName == _EMPTY_ {
+		// if consumer name is empty, use the legacy ephemeral endpoint
+		ccSubj = fmt.Sprintf(apiLegacyConsumerCreateT, stream)
+	} else if err := checkConsumerName(consumerName); err != nil {
+		return nil, err
+	} else if js.nc.serverMinVersion(2, 9, 0) {
+		// if above server version 2.9.0, use the endpoints with consumer name
```

Comment:
If not checking the server name, what I think should be done is just in checking if the new config Name is specified. If it is, then that signals the intent to use the new API endpoints. If user specifies Name, but connects to an older server, they will get a timeout: this is not ideal, but I don't think checking server version is reliable anyway (could be connected to one 2.9.0, but server accepting the request be older version, or vice versa).

Reply:
Makes sense, I'll change it.

Reply:
So, just to clarify - when user provides only `Durable`, we should use the old endpoints (`CONSUMER.DURABLE.CREATE`)?
If so, that makes it problematic for e.g. `Subscribe("foo", nats.Durable("cons"))`, as basically it will always use old API subject, unless we add another `SubOpt` for just setting consumer name.

Reply:
We should then possibly set the Name to the Durable before calling js.AddConsumer(). But then, I agree that we have the situation where this is not a user choice and so connecting to older server would be a problem...

Reply:
So from what I understand, we have following options:
1. Not verify the server version compatibility and call the new API if `Name` is set - problem here is, `Subscribe()` will always call the old API when combined with `Bind()`/`Durable()` options
2. Do as above, but set `Name` to the value of `Durable` in `Subscribe()` - that means however that we would always be using new API, even for older servers (that's not an option)
3. Verify server version when choosing the right subject in `upsertConsumer()` - here, we might have an issue of connecting to an older server, as you described in the original comment.

Looking at those, I would still lean towards option 3 - I could strip version check in `AddConsumer()`/`UpdateConsumer()`, but leave the version check when choosing the right subject in `upsertConsuper()`. What do you think?

Reply:
If we want to start testing server compatibility across client versions maybe we could add some additional tests with a different go module to run the older server with new client, something like this maybe: https://github.com/nats-io/nats.go/commit/55ccb7b1f2193e0e9febbf8a23595705b4b7e4a5?diff=unified

Reply:
@aricart and I both check the server version so we can use the new endpoints if the user sets only durable.

Reply:
@wallyqs I like the idea, but I'm a bit whether it won't introduce a horrible mess in the future, with each breaking change on the server requiring us to create separate `go.mod` files to point to the right server version. 

I suppose it might be a good idea to have e.g. a separate repository running clients tests on multiple server versions. That's something we briefly discusses with @Jarema some time ago, we can revisit this.

Reply:
It would not have been much of an issue if we were passing the ConsumerConfig as an option (instead of individual options that update a ConsumerConfig, although the advantage of options the way we do is that we *could* (but not really do) handle "set" vs "not set" for boolean properties). If the ConsumerConfig was provided, then it would be up to the user to set the ConsumerConfig.Name or not, and that would have been the only driver to select new API or not. Of course, we could add a ConsumerName() SubOpt to do that.

Since the rest of the group seem to be using the server version to send to new API for a durable which consumer config's name is not set, we could do that (and I will update the C client to match that behavior).

Reply:
`ConsumerName()` is an option for sure, but then we would also have to add some alternative version of `Bind()` which sets stream and name, rather than stream and durable (figuring out a name for this option would be fun for sure).

Reply:
Go with the server version check then if all agree that this is the best approach we have.

Reply:
We can still make it possible if needed to pass the ConsumerConfig to configure Subscribe without options by implementing the SubOpt interface, that would avoid having to keep up with ConsumerConfig latest changes.  Something like this would work:

```go
func (config *ConsumerConfig) configureSubscribe(opts *subOpts) error {
// Implements SubOpt interface...
}

// Can use ConsumerConfig struct to configure without extra options.
_, err := js.Subscribe("foo", cb, &nats.ConsumerConfig{
	Durable: "bar",
	Replicas: 1,
})
```

Reply:
It's a nice option to have (we can consider adding it of course), but I still believe picking the right subject based on server version is useful - if we add it, then the new API will be used by the client as soon as both server and client are updated to the new version (transparently, without needing to change the application code).

Reply:
The pure version orientated approach means users do not get a chance to say they need time to update infrastructure like ACLs and so forth to start using it.

Reply:
Hm, not sure how big of an issue that would be vs forcing the user to intentionally start using the new API by changing their application code (IMO this transition should ideally be seamless, as the user of client library I don't need to care about the API subjects). What do you think?

Reply:
There is a class of user who care deeply to lock down the APIs according to needs. Also a class of user who have spent considerable time in navigating the API subjects for cross account use via imports and exports. 

A major design feature of these API subjects is to enable that lock down. Or to be selectively imported to manage permissions and restrictions. 

The Venn diagrams of users likely to pay for NATS and those who care deeply for subject security probsbly has quite a lot of overlap. 

So you might not care, but I think we should consider if introducing new features in the most user hostile way possible is perhaps not the right thing - especially considering the users most likely affected.

Reply:
The issue with the ACL is possibly real, but in reality, the same logic can be applied the other way. If they deploy the new client but don't update the servers, they can update their ACLs and then deploy the servers.

The client possibly could have a way of rejecting the use of the new API (I do for test purposes). But at some point, the clients become too complex and too flexible.

Reply:
The question is how do users know? Do we have effective communication channels to tell people about these changes? Release notes and blog post are notoriously ineffective - especially as ours tend to be enormous. 

So do we feel users are served well enough by the communication and warning we give them about upcoming changes?

In another world these would be considered  breaking changes and tooling and just human behaviour is aware of major changes. These are not just new features. They majorly change existing code simply because it happens to point at another version server.

As much as I loathe the go major version change behaviour this does demonstrate the utility of that.